diff -uNr linux-3.9.6/drivers/net/ethernet/intel/e1000/e1000_hw.h linux-3.9.6-ilab.t.vlan.transparency/drivers/net/ethernet/intel/e1000/e1000_hw.h
--- linux-3.9.6/drivers/net/ethernet/intel/e1000/e1000_hw.h	2013-06-13 19:52:08.000000000 +0200
+++ linux-3.9.6-ilab.t.vlan.transparency/drivers/net/ethernet/intel/e1000/e1000_hw.h	2013-06-20 11:11:06.000000000 +0200
@@ -478,7 +478,7 @@
 #define VLAN_TAG_SIZE  4	/* 802.3ac tag (not DMAed) */
 
 /* Ethertype field values */
-#define ETHERNET_IEEE_VLAN_TYPE 0x8100	/* 802.3ac packet */
+#define ETHERNET_IEEE_VLAN_TYPE ETH_P_8021Q	/* 802.3ac packet */
 #define ETHERNET_IP_TYPE        0x0800	/* IP packets */
 #define ETHERNET_ARP_TYPE       0x0806	/* Address Resolution Protocol (ARP) */
 
diff -uNr linux-3.9.6/drivers/net/ethernet/intel/igb/e1000_defines.h linux-3.9.6-ilab.t.vlan.transparency/drivers/net/ethernet/intel/igb/e1000_defines.h
--- linux-3.9.6/drivers/net/ethernet/intel/igb/e1000_defines.h	2013-06-13 19:52:08.000000000 +0200
+++ linux-3.9.6-ilab.t.vlan.transparency/drivers/net/ethernet/intel/igb/e1000_defines.h	2013-06-20 11:12:05.000000000 +0200
@@ -344,7 +344,7 @@
 #define E1000_COLD_SHIFT                12
 
 /* Ethertype field values */
-#define ETHERNET_IEEE_VLAN_TYPE 0x8100  /* 802.3ac packet */
+#define ETHERNET_IEEE_VLAN_TYPE ETH_P_8021Q  /* 802.3ac packet */
 
 #define MAX_JUMBO_FRAME_SIZE    0x3F00
 
diff -uNr linux-3.9.6/drivers/net/ethernet/intel/ixgbe/ixgbe_type.h linux-3.9.6-ilab.t.vlan.transparency/drivers/net/ethernet/intel/ixgbe/ixgbe_type.h
--- linux-3.9.6/drivers/net/ethernet/intel/ixgbe/ixgbe_type.h	2013-06-13 19:52:08.000000000 +0200
+++ linux-3.9.6-ilab.t.vlan.transparency/drivers/net/ethernet/intel/ixgbe/ixgbe_type.h	2013-06-20 11:13:48.000000000 +0200
@@ -1490,7 +1490,7 @@
 #define IXGBE_VMVIR_VLANA_DEFAULT 0x40000000 /* Always use default VLAN */
 #define IXGBE_VMVIR_VLANA_NEVER   0x80000000 /* Never insert VLAN tag */
 
-#define IXGBE_ETHERNET_IEEE_VLAN_TYPE 0x8100  /* 802.1q protocol */
+#define IXGBE_ETHERNET_IEEE_VLAN_TYPE ETH_P_8021Q  /* 802.1q protocol */
 
 /* STATUS Bit Masks */
 #define IXGBE_STATUS_LAN_ID         0x0000000C /* LAN ID */
diff -uNr linux-3.9.6/include/uapi/linux/if_ether.h linux-3.9.6-ilab.t.vlan.transparency/include/uapi/linux/if_ether.h
--- linux-3.9.6/include/uapi/linux/if_ether.h	2013-06-13 19:52:08.000000000 +0200
+++ linux-3.9.6-ilab.t.vlan.transparency/include/uapi/linux/if_ether.h	2013-06-20 11:09:17.000000000 +0200
@@ -61,7 +61,7 @@
 #define ETH_P_RARP      0x8035		/* Reverse Addr Res packet	*/
 #define ETH_P_ATALK	0x809B		/* Appletalk DDP		*/
 #define ETH_P_AARP	0x80F3		/* Appletalk AARP		*/
-#define ETH_P_8021Q	0x8100          /* 802.1Q VLAN Extended Header  */
+#define ETH_P_8021Q	0x7100          /* iLab.t patched 802.1Q VLAN Extended Header  */
 #define ETH_P_IPX	0x8137		/* IPX over DIX			*/
 #define ETH_P_IPV6	0x86DD		/* IPv6 over bluebook		*/
 #define ETH_P_PAUSE	0x8808		/* IEEE Pause frames. See 802.3 31B */
